### PR TITLE
[AutoTP] Add support for vocab-parallel LM head and related configurations

### PR DESCRIPTION
**Pull request by jinyouzhi**
## Motivation

* Address PR-B in https://github.com/deepspeedai/DeepSpeed/issues/8173

## Changes

This pull request adds support for vocabulary-parallel language modeling (LM) heads in DeepSpeed's tensor parallelism (TP) system. It introduces a new `VocabParallelLinear` layer that keeps logits sharded across TP ranks for untied LM heads and installs a compatible distributed loss automatically. The changes ensure correct partitioning, validation, and integration with the rest of the DeepSpeed engine and configuration system.

**Key changes include:**

### Vocabulary-parallel LM head support

* Introduced a `vocab_parallel_lm_head` configuration option in `TPTrainingConfig` and passed it through the relevant APIs to enable this feature. [[1]](diffhunk://#diff-4c647d18231cb189dcebad994b456ea76c6ccc9d1771a581c5b5f03a0f080527R53-R55) [[2]](diffhunk://#diff-3e2a27cf6940a22f2abf4c9e19d007d3824b6a97afb8d8cc028c7efa9ddd7e44R210) [[3]](diffhunk://#diff-cf74152f0933aa49e65fc12cb0017a40f5b6e2880bd7c01e0d2b26693b6c88e7R302) [[4]](diffhunk://#diff-e6635a81d2c2bf0938b5f83b1c4945e0f344e0106191a6960df8ee4aa64cc55fR874-R881) [[5]](diffhunk://#diff-e6635a81d2c2bf0938b5f83b1c4945e0f344e0106191a6960df8ee4aa64cc55fR914-R922)
* Added logic in `auto_tp.py` to detect and correctly replace untied LM heads with the new `VocabParallelLinear` layer, along with validation to ensure the vocab head is not tied to the embedding. [[1]](diffhunk://#diff-3e2a27cf6940a22f2abf4c9e19d007d3824b6a97afb8d8cc028c7efa9ddd7e44R229) [[2]](diffhunk://#diff-3e2a27cf6940a22f2abf4c9e19d007d3824b6a97afb8d8cc028c7efa9ddd7e44R377-R379) [[3]](diffhunk://#diff-3e2a27cf6940a22f2abf4c9e19d007d3824b6a97afb8d8cc028c7efa9ddd7e44R444-R446) [[4]](diffhunk://#diff-3e2a27cf6940a22f2abf4c9e19d007d3824b6a97afb8d8cc028c7efa9ddd7e44R519-R543)

### New layer and loss integration

* Implemented the `VocabParallelLinear` layer in `layers.py`, which ensures logits remain sharded and provides partitioning metadata. [[1]](diffhunk://#diff-0df5862062e8a6030b3b4c689914f06701a7023014f94d353cbd756e8a898776L30-R30) [[2]](diffhunk://#diff-0df5862062e8a6030b3b4c689914f06701a7023014f94d353cbd756e8a898776R963-R978)
* Updated the DeepSpeed engine to automatically install a distributed causal LM loss compatible with vocab-parallel heads, and to handle model metadata/validation accordingly. [[1]](diffhunk://#diff-e6635a81d2c2bf0938b5f83b1c4945e0f344e0106191a6960df8ee4aa64cc55fR849-R864) [[2]](diffhunk://#diff-e6635a81d2c2bf0938b5f83b1c4945e0f344e0106191a6960df8ee4aa64cc55fL924-R938)

### Miscellaneous

* Made minor logic corrections and refactoring to ensure correct replacement heuristics and fallback configuration. [[1]](diffhunk://#diff-3e2a27cf6940a22f2abf4c9e19d007d3824b6a97afb8d8cc028c7efa9ddd7e44L495-R503) [[2]](diffhunk://#diff-3e2a27cf6940a22f2abf4c9e19d007d3824b6a97afb8d8cc028c7efa9ddd7e44L727-R760)
* Exposed new loss functions and configuration utilities in the `deepspeed.sequence` module's public API.

These changes collectively enable efficient, distributed training of models with untied, sharded vocabulary projection heads and ensure compatibility with DeepSpeed's tensor parallelism ecosystem.

## Limitations
*  TP x SP mesh are unsupported.

### CODE REVIEW

**Comment by delock** [comment]
Hi @jinyouzhi when this PR is ready for review just ping me, thanks!

File: deepspeed/runtime/tensor_parallel/config.py [hunk 1]
```diff
@@ -50,6 +50,9 @@ class TPTrainingConfig(DeepSpeedConfigModel):
     tp_overlap_comm: bool = False
     """ Whether to overlap communication with computation. Currently, only allreduce supports overlap. """
 
+    vocab_parallel_lm_head: bool = False
```
[the remark this answers is not in the document]
Comment (jinyouzhi) [reply deepspeed/runtime/tensor_parallel/config.py:53]: DCO passed.

File: deepspeed/runtime/engine.py [hunk 2]
```diff
@@ -921,8 +935,7 @@ def _apply_autotp_partitioning(self, model, tp_config):
             tp_config.injection_policy_tuple = injection_policy
             replace_transformer_layer(client_module, model, None, tp_config, model_config, training_mode=True)
 
-        setattr(model, UNIVERSAL_CHECKPOINT_INFO, collect_autotp_universal_checkpoint_info(model))
-        setattr(model, "ds_autotp_parsed", True)
+        finalize_autotp(attach_uc_metadata=True)
```
[the remark this answers is not in the document]
Comment (jinyouzhi) [reply deepspeed/runtime/engine.py:956]: Fixed in  [a3b440523 ](https://github.com/deepspeedai/DeepSpeed/pull/8309/commits/a3b4405236c60a7830b6889d328032a83516e2b8). The heuristic AutoTP path now creates a dedicated  AutoTP  instance for the output head and explicitly replaces the single supported  `lm_head` / `embed_out`  before finalization. It also fails clearly when no supported head or multiple candidate heads are found, instead of silently leaving the full head replicated. I added a 2-rank end-to-end regression using a real Llama model through  `deepspeed.initialize` , forward, distributed causal-LM loss, and backward.

File: deepspeed/module_inject/auto_tp.py [hunk 3]
```diff
@@ -508,6 +516,31 @@ def _create_column_parallel_layer(self, module, spec: TPLayerSpec, name: str):
...
+        log_dist(
+            f"AutoTP: vocab_parallel_lm_head keeps '{name}' vocabulary-sharded and installs the "
+            f"distributed causal-LM loss",
+            ranks=[0])
+        return VocabParallelLinear(child, self.mp_group, name=name, tp_meta=self.tp_meta)
+
+    def _validate_untied_vocab_head(self, lm_head):
+        for _, module in self.module.named_modules():
+            if isinstance(module, nn.Embedding) and getattr(module, "weight", None) is lm_head.weight:
+                raise ValueError("A no-gather vocab-parallel LM head requires untied embedding and output weights")
```
[the remark this answers is not in the document]
Comment (jinyouzhi) [reply deepspeed/module_inject/auto_tp.py:556]: Fixed in  [a3b440523 ](https://github.com/deepspeedai/DeepSpeed/pull/8309/commits/a3b4405236c60a7830b6889d328032a83516e2b8).  AutoTP  now records the identities of vocab-head modules tied to embeddings when it is constructed, before traversal can replace an embedding parameter.  `_validate_untied_vocab_head()`  checks this original tie information as well as the current parameter identity. A regression test covers an  `embedding_rowwise`  plan followed by a  `colwise`  tied  `lm_head`  and verifies that the configuration is rejected before weight sharing can be broken.

**Comment by jinyouzhi** [comment]
> Hi @jinyouzhi when this PR is ready for review just ping me, thanks!

Hi, this PR is ready for review now. Thank you for your patience! @delock 

During development, I identified and fixed some issues related to SP in a separate PR: #8457.

**Comment by delock** [comment]
Hi @jinyouzhi Thanks for the PR — the loss kernel and backward-compat handling for `vocab_sequence_parallel_cross_entropy` look solid. Two changes requested, one suggestion:

**1. License header (new file)**
`tests/unit/v1/sequence_parallelism/test_vocab_parallel_cross_entropy.py` uses the `Copyright (c) Microsoft Corporation` header. New files should use the plain header only:
```
# SPDX-License-Identifier: Apache-2.0
# DeepSpeed Team
```

**2. `vocab_parallel_lm_head` silently overrides explicit lm_head specs**
In `_replace_with_config`, the vocab-head check runs before spec lookup, so with the flag on, an explicit `colwise_gather_output` (or even `rowwise`) spec for `lm_head` in `partition_config` is silently ignored. Please raise (or at least warn) on this conflict instead, and add a test pinning the behavior.

**Suggestion (non-blocking)**: a minimal save → load → forward-loss parity test for the vocab-sharded head would be good to have, since checkpoint paths have no coverage for the new `VocabParallelLinear` layout.

**Comment by jinyouzhi** [comment]
@delock Thank you for your detailed review! I have updated the PR according to your comments and guidance. Please take another look when you have a chance.

**Comment by delock** [comment]
**LM-head name detection is duplicated in several coupled places — should all reuse `lm_head_patterns`**

The `("lm_head", "embed_out")` check is duplicated in `_create_column_parallel_layer`, `_create_row_parallel_layer`, `_configure_gathered_column_tie_fallbacks`, and the new `_is_lm_head_name`. These sites are coupled — e.g. the vocab-head skip in `_configure_gathered_column_tie_fallbacks` only stays correct if its notion of "LM head" matches the legacy checks — and for coupled duplication the right fix is reuse, not a second hardcoded copy. Please make these call sites resolve the name set from `lm_head_patterns` (with the current pair as the default) so there's a single source of truth.

**Comment by delock** [comment]
`_vocab_metadata_cache` holds strong references to process groups — in same-process train+inference setups (RL) where groups are rebuilt, this pins them and their NCCL resources for the life of the process. Suggest `WeakKeyDictionary` keyed by group to avoid the leak.

**Comment by jinyouzhi** [comment]
> `_vocab_metadata_cache` holds strong references to process groups — in same-process train+inference setups (RL) where groups are rebuilt, this pins them and their NCCL resources for the life of the process. Suggest `WeakKeyDictionary` keyed by group to avoid the leak.

Good catch. I removed the reference to the `tp_group` object from the cache entirely, rather than keeping it as a key — `tp_group` is now only held by `VocabParallelCausalLMLoss` itself (as `self.tp_group`), and the cache just stores the resolved `start/end_idx`.

I think this is safe enough for the multi-model/multi-shard/multi-loss case, since each loss instance owns its own `tp_group` and cache with matching lifetimes. Curious how this compares to `WeakRef` in your view — let me know if you'd still prefer that approach.

**Review by delock — APPROVED** [review]

**Comment by delock** [comment]
@jinyouzhi can you fix CI failure?  Thanks!

**Comment by jinyouzhi** [comment]
> @jinyouzhi can you fix CI failure?  Thanks!

OK sure
